pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -0,0 +1,235 @@
+        let sleep_until = start + task_config.period;
+        tokio::select! {
+            _ = tokio::time::sleep_until(sleep_until) => {},
+            _ = cancel.cancelled() => {
+                info!("shutting down");
+                return;
+            }
+        }
+    }
+}
+
+enum IterationOutcome {
+    NoPressure,
+    Cancelled,
+    Finished,
+}
+
+async fn disk_usage_eviction_iteration(
+    task_config: &DiskUsageEvictionTaskConfig,
+    tenants_dir_fd: &mut sync_wrapper::SyncWrapper<Dir>,
+    cancel: CancellationToken,
+) -> anyhow::Result<IterationOutcome> {
+    let stat: Statvfs = statvfs::fstatvfs(tenants_dir_fd.get_mut())
+        .context("statvfs failed, presumably directory got unlinked")?;
+
+    // https://unix.stackexchange.com/a/703650
+    let blocksize = if stat.fragment_size() > 0 {
+        stat.fragment_size()
+    } else {
+        stat.block_size()
+    };
+
+    // use blocks_available (b_avail) since, pageserver runs as unprivileged user
+    let avail_bytes = stat.blocks_available() * blocksize;
+    let total_bytes = stat.blocks() * blocksize;
+
+    // At any reasonable filesystem size, the rounding-down error caused by this integer division is negligible.
+    let usage_pct = 100 - ((100 * (avail_bytes as u64)) / (100 * (total_bytes as u64)));
+
+    let pressure = [
+        ("min_avail_bytes", avail_bytes < task_config.min_avail_bytes),
+        ("max_usage_pct", usage_pct > task_config.max_usage_pct),
+    ];
+
+    debug!(avail_bytes, total_bytes, usage_pct, ?pressure, "disk usage");
+
+    if !pressure.iter().any(|(_, problem)| *problem) {
+        return Ok(IterationOutcome::NoPressure);
+    }
+
+    warn!(
+        avail_bytes,
+        total_bytes,
+        usage_pct,
+        ?pressure,
+        "running disk usage based eviction"
+    );
+
+    // get a snapshot of the list of tenants
+    let mut data: Vec<(TenantId, Vec<(TimelineId, LocalLayerInfo)>)> =
+        crate::tenant::mgr::list_tenants()
+            .await
+            .context("get list of tenants")?
+            .into_iter()
+            .map(|(tenant_id, _)| (tenant_id, Vec::new()))
+            .collect();
+
+    // get layer infos for all timelines for all tenants
+    for (tenant_id, layer_infos) in &mut data {
+        let tenant = match crate::tenant::mgr::get_tenant(*tenant_id, false).await {
+            Ok(tenant) => tenant,
+            // this can happen if tenant has lifecycle transition after we fetched it above
+            Err(e) => {
+                warn!("failed to get tenant {tenant_id}: {e:#}");
+                continue;
+            }
+        };
+        if cancel.is_cancelled() {
+            return Ok(IterationOutcome::Cancelled);
+        }
+
+        let timelines = tenant.list_timelines();
+        for tl in timelines {
+            layer_infos.extend(
+                tl.get_local_layer_infos()
+                    .into_iter()
+                    .map(|layer_infos| (tl.timeline_id, layer_infos)),
```

Comment:
Letting go of the `Arc<Timeline>` will require hashmap lookups in the next phase, would reconsider that.

Reply:
In a future world where active 'Arc<Timeline>` / `Arc<Tenant>` holds up lifecycle operations, I wouldn't want this background loop to do that.
Then again, my latest force push adds a caching hashmap for the tenants & timelines where we did need to do evictions.
It's better than holding _all_ tenants & timelines upfront, but yeah, I see your point.

Please take a look at the new code and open a new conversation there.

Reply:
Oki maybe I understand now, we want the ids for sorting.

Reply:
This has now been rewritten not to worry about arcs. We also looked into not introducing a detach/delete race with evicting, so resolving now.